### PR TITLE
docs: added Core Services section

### PR DESCRIPTION
**Pull request by marcellorinaldo**
Signed-off-by: Marcello Martina <martina.marcello.rinaldo@outlook.com>

> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

This PR ports the old "Core services" section documentation to the new one.

**Related Issue:** N/A.

**Description of the solution adopted:** N/A.

**Screenshots:** N/A.

**Any side note on the changes made:** N/A.

### CODE REVIEW

**Comment by marcellorinaldo** [comment]
@mattdibi can you please review?

File: docs/core-services/command-service.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Command Service
+
+The Command Service provides methods for running system commands from the Kura web console or from Kapua. In the Kura web console, the service is available clicking on the **Command** tab under the **Device** section, while for the cloud platform please refer to the official documentation.
+
+To run a command simply fill the **Execute** field with the command and click the **Execute** button. The service also provides the ability for a script to execute using the **File** option of the **Command** tab in the Kura web console or the Kapua Cloud Console. This script must be compressed into a zip file with the eventual, associated resource files.
+
+Once the file is selected and **Execute** is clicked, the zip file is sent embedded in an MQTT message on the device. The Command Service in the device stores the file in /tmp, unzips it, and tries to execute a shell script if one is present in the file. Note that in this case, the Execute parameter cannot be empty; a simple command, such as `ls -l /tmp`, may be entered.
+
+!!! warn
```
Comment (mattdibi) [inline docs/core-services/command-service.md]:
```suggestion
!!! warning
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 2]
```diff
@@ -0,0 +1,1278 @@
...
+    "org.eclipse.kura.internal.rest.provider.RestService",
+    "com.eurotech.framework.reboot.RebootService"
+]
+```
+
+### Lists all the component configurations of all the ConfigurableComponents
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations`
+**Response**:
+```JSON
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]: Fix formatting.

```suggestion
**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations`

**Response**:
```JSON
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 3]
```diff
@@ -0,0 +1,1278 @@
...
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/factoryComponents/{pid}?takeSnapshot={takeSnapshot}`
+
+- **takeSnapshot**: can be either true or false. If set to true, after the deletion, the framework will take a snapshot. 
+
+### List the tracked configurable component Pids
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents`
+**Response**:
+```JSON
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]:
```suggestion
**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents`

**Response**:
```JSON
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 4]
```diff
@@ -0,0 +1,1278 @@
...
+    },
+  ...
+]
+```
+
+### List the configurations of all the ConfigurableComponents that match a filter
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/byFilter/(service.pid=org.eclipse.kura.clock.ClockService)`
+**Response**:
+```JSON
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]:
```suggestion
**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/byFilter/(service.pid=org.eclipse.kura.clock.ClockService)`

**Response**:
```JSON
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 5]
```diff
@@ -0,0 +1,1278 @@
...
+        }
+    }
+]
+```
+
+### Get the configuration of the ConfigurableComponent matching a pid
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/byPid/org.eclipse.kura.clock.ClockService`
+**Response**:
+```JSON
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]:
```suggestion
**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/byPid/org.eclipse.kura.clock.ClockService`

**Response**:
```JSON
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 6]
```diff
@@ -0,0 +1,1278 @@
...
+        }
+    }
+}
+```
+
+### Get the default configuration of the ConfigurableComponent matching a pid
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/byPid/org.eclipse.kura.clock.ClockService/_default`
+**Response**:
+```JSON
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]:
```suggestion
**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/byPid/org.eclipse.kura.clock.ClockService/_default`

**Response**:
```JSON
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 7]
```diff
@@ -0,0 +1,1278 @@
...
+        }
+    }
+}
+```
+
+### Update the component configuration identified matching a pid
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/byPid/org.eclipse.kura.clock.ClockService/_update`
+**Request body**:
+```JSON
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]:
```suggestion
**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/byPid/org.eclipse.kura.clock.ClockService/_update`

**Request body**:
```JSON
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 8]
```diff
@@ -0,0 +1,1278 @@
...
+            }
+        }
+    }
+    ```
+
+### Update the configuration of multiple configurable components
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/_update`
+**Request body**:
+```JSON
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]:
```suggestion
**Request**: URL - `https://<gateway-ip>/services/configuration/v1/configurableComponents/configurations/_update`

**Request body**:
```JSON
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 9]
```diff
@@ -0,0 +1,1278 @@
...
+        }
+    ]
+}
+```
+
+### List all the available snapshot IDs
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/snapshots`
+**Response**:
+```JSON
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]:
```suggestion
**Request**: URL - `https://<gateway-ip>/services/configuration/v1/snapshots`

**Response**:
```JSON
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 10]
```diff
@@ -0,0 +1,1278 @@
...
+    1630930797402,
+    1630930805305
+]
+```
+
+### Get the content of a given snapshot
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/snapshots/{id}`
+**Response**:
+```JSON
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]:
```suggestion
**Request**: URL - `https://<gateway-ip>/services/configuration/v1/snapshots/{id}`

**Response**:
```JSON
```

File: docs/core-services/configuration-service-rest-v1.md [hunk 11]
```diff
@@ -0,0 +1,1278 @@
...
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/snapshots/_rollback`
+**Response**:
+```JSON
+1631093011618
+```
+
+### Upload a snapshot as XML
+
+**Request**: URL - `https://<gateway-ip>/services/configuration/v1/snapshots/_upload`
+**Request body**:
```
Comment (mattdibi) [inline docs/core-services/configuration-service-rest-v1.md]:
```suggestion
### Trigger the framework to take and persist a snapshot

**Request**: URL - `https://<gateway-ip>/services/configuration/v1/snapshots/_write`

**Response**:
```JSON
1631095409516
```

### Rollbacks to the snapshot identified by the provided ID

**Request**: URL - `https://<gateway-ip>/services/configuration/v1/snapshots/_rollback`

**Response**:
```JSON
1631093011618
```

### Upload a snapshot as XML

**Request**: URL - `https://<gateway-ip>/services/configuration/v1/snapshots/_upload`

**Request body**:
```

File: docs/core-services/rest-service.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+By default, REST service providers register their services using the context path ```/services```. The REST service provides the **BASIC** Authentication support and HTTPS client certificate authentication support.
+
+REST API access is available on all HTTP ports defined in the [HTTP/HTTPS Configuration](doc:httphttps-configuration) section, unless access is restricted to dedicated ports using the corresponding configuration parameter (see below).
+
+Certificate authentication support is only available on the **HTTPS With Certificate Authentication Ports** configured in [HTTP/HTTPS Configuration](doc:httphttps-configuration) section.
+
+Kura Identity names and passwords can be used for **BASIC** Authentication. Certificate authentication follows the same rules as [Gateway Administration Console Authentication](doc:security-configuration).
+
+!!! warn
```
Comment (mattdibi) [inline docs/core-services/rest-service.md]:
```suggestion
!!! warning
```

File: docs/core-services/rest-service.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
+# REST Service
+
+Kura provides a built-in REST Service based on the [osgi-jax-rs-connector](https://github.com/hstaudacher/osgi-jax-rs-connector) project.
+
+By default, REST service providers register their services using the context path ```/services```. The REST service provides the **BASIC** Authentication support and HTTPS client certificate authentication support.
```
Comment (mattdibi) [inline docs/core-services/rest-service.md]:
```suggestion
By default, REST service providers register their services using the context path `/services`. The REST service provides the **BASIC** Authentication support and HTTPS client certificate authentication support.
```

File: docs/core-services/active-mq-artemis-broker.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+![ActiveMQ Artemis MQTT Broker](./images/active-mq-artemis-broker.png)
+
+This service exposes the following configuration parameters:
+
+- **Enabled** - (Required) - Enables the broker instance
+- **Broker XML** - Broker XML configuration. An empty broker configuration will disable the broker.
+- **Required protocols** - A comma seperated list of all required protocol factories (e.g. AMQP or MQTT)
+- **User configuration** - (Required) - User configuration in the format: user=password|role1,role2,...
+- **Default user name*** - The name of the default user
```
Comment (mattdibi) [inline docs/core-services/active-mq-artemis-broker.md]:
```suggestion
- **Default user name** - The name of the default user
```

File: docs/core-services/container-orchestration-provider-usage.md [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+
+After specifying container parameters, ensure to set **Enabled** to **true** and press **Apply**. The container engine will then pull the respective image, spin up and start the container. If the gateway or the framework is power cycled, and the container and Container Orchestration Service are set to **enabled**, the framework will automatically start the container again upon startup.
+
+![Container Orchestration Provider Container Configuration](./images/container-orchestration-provider-container-configuration.png)
+
+
+
+## Stopping the container
+
+!!! warn
```
Comment (mattdibi) [inline docs/core-services/container-orchestration-provider-usage.md]:
```suggestion
!!! warning
```

File: docs/core-services/nvidia-triton-server-inference-engine.md [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+
+Before you can use the Triton Docker image you must install [Docker](https://docs.docker.com/engine/install). If you plan on using a GPU for inference you must also install the [NVIDIA Container Toolkit](https://github.com/NVIDIA/nvidia-docker).
+
+Pull the image using the following command.
+
+```
+$ docker pull nvcr.io/nvidia/tritonserver:<xx.yy>-py3
+```
+
+Where \<xx.yy\> is the version of Triton that you want to pull.
```
Comment (mattdibi) [inline docs/core-services/nvidia-triton-server-inference-engine.md]: This doesn't get rendered correctly. Try:

```suggestion
Where `<xx.yy>` is the version of Triton that you want to pull.
```

File: docs/core-services/nvidia-triton-server-inference-engine.md [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+- **Inference Models**: a comma-separated list of inference model names that the server will load. The models have to be already present in the filesystem where the server is running. This option simply tells the server to load the given models from a local or remote repository.
+- **Timeout (in seconds) for time consuming tasks**: Timeout (in seconds) for time consuming tasks like server startup, shutdown or model load. If the task exceeds the timeout, the operation will be terminated with an error.
+- **Max. GRPC message size (bytes)**: this field controls the maximum allowed size for the GRPC calls to the server instance. By default, size of 4194304 bytes (= 4.19 MB) is used. Increase this value to be able to send large amounts of data as input to the Triton server (like Full HD images). The Kura logs will show the following error when exceeding such limit:
+    ```
+    io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: gRPC message exceeds maximum size 4194304
+    ```
+
+> **Note**: Pay attention on the ports used for communicating with the Triton Server. The default ports are the 8000-8002, but these are tipically used by Kura for debug purposes.
```
Comment (mattdibi) [inline docs/core-services/nvidia-triton-server-inference-engine.md]:
```suggestion
!!! note
    Pay attention on the ports used for communicating with the Triton Server. The default ports are the 8000-8002, but these are tipically used by Kura for debug purposes.
```

File: docs/core-services/nvidia-triton-server-inference-engine.md [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+- **Inference Models**: a comma-separated list of inference model names that the server will load. The models have to be already present in the filesystem where the server is running. This option simply tells the server to load the given models from a local or remote repository.
+- **Local backends path**: Specify the path on the filesystem where the backends are stored.
+- **Optional configuration for the local backends**: A semi-colon separated list of configuration for the backends. i.e. tensorflow,version=2;tensorflow,allow-soft-placement=false 
+- **Timeout (in seconds) for time consuming tasks**: Timeout (in seconds) for time consuming tasks like server startup, shutdown or model load. If the task exceeds the timeout, the operation will be terminated with an error.
+- **Max. GRPC message size (bytes)**: this field controls the maximum allowed size for the GRPC calls to the server instance.
+
+> **Note**: Pay attention on the ports used for communicating with the Triton Server. The default ports are the 8000-8002, but these are tipically used by Kura for debug purposes.
```
Comment (mattdibi) [inline docs/core-services/nvidia-triton-server-inference-engine.md]:
```suggestion
!!! note
    Pay attention on the ports used for communicating with the Triton Server. The default ports are the 8000-8002, but these are tipically used by Kura for debug purposes.
```

File: docs/core-services/nvidia-triton-server-inference-engine.md [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+- **Inference Models**: a comma-separated list of inference model names that the server will load. The models have to be already present in the filesystem where the server is running. This option simply tells the server to load the given models from a local or remote repository.
+- **Local backends path**: Only for a local instance, specify the path on the filesystem where the backends are stored.
+- **Optional configuration for the local backends**: Only for local instance, a semi-colon separated list of configuration for the backends. i.e. tensorflow,version=2;tensorflow,allow-soft-placement=false 
+- **Timeout (in seconds) for time consuming tasks**: Timeout (in seconds) for time consuming tasks like server startup, shutdown or model load. If the task exceeds the timeout, the operation will be terminated with an error.
+- **Max. GRPC message size (bytes)**: this field controls the maximum allowed size for the GRPC calls to the server instance.
+
+> **Note**: Pay attention on the ports used for communicating with the Triton Server. The default ports are the 8000-8002, but these are tipically used by Kura for debug purposes.
```
Comment (mattdibi) [inline docs/core-services/nvidia-triton-server-inference-engine.md]:
```suggestion
!!! note
    Pay attention on the ports used for communicating with the Triton Server. The default ports are the 8000-8002, but these are tipically used by Kura for debug purposes.
```

File: docs/core-services/device-configuration-changes.md [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+The configuration change manager allows to collection of the PIDs of the components that have changed in the system. The configurable properties of this component are:
+
+- **Enable**: whether to enable or not this component.
+- **CloudPublisher Target Filter**: allows to specify the cloud publisher to use for sending the produced messages.
+- **Notification send delay (sec)**: allows to stack the changed PIDs for a given time before sending. In other words, it allows to group together sequences of PIDs that are changed in that time frame. This is to prevent message flooding at reboot or rollback.
+
+The collected PIDs are sent to the Cloud Publisher as a `KuraMessage` with a payload body in JSON format. The timestamp of the `KuraMessage` is set to the last detected configuration change event. An example of message body is:
+
+```JSON
```
Comment (mattdibi) [inline docs/core-services/device-configuration-changes.md:23]:
```suggestion
```json
```

File: docs/core-services/rest-service.md [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+
+REST API access is available on all HTTP ports defined in the [HTTP/HTTPS Configuration](doc:httphttps-configuration) section, unless access is restricted to dedicated ports using the corresponding configuration parameter (see below).
```
Comment (mattdibi) [inline docs/core-services/rest-service.md:7]: Broken link, is this expected?

File: docs/core-services/rest-service.md [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+
+Certificate authentication support is only available on the **HTTPS With Certificate Authentication Ports** configured in [HTTP/HTTPS Configuration](doc:httphttps-configuration) section.
+
+Kura Identity names and passwords can be used for **BASIC** Authentication. Certificate authentication follows the same rules as [Gateway Administration Console Authentication](doc:security-configuration).
```
Comment (mattdibi) [inline docs/core-services/rest-service.md]: Broken link, is this expected?

File: docs/core-services/rest-service.md [hunk 23]
```diff
@@ -0,0 +1,46 @@
+# REST Service
+
+Kura provides a built-in REST Service based on the [osgi-jax-rs-connector](https://github.com/hstaudacher/osgi-jax-rs-connector) project.
+
+By default, REST service providers register their services using the context path ```/services```. The REST service provides the **BASIC** Authentication support and HTTPS client certificate authentication support.
+
+REST API access is available on all HTTP ports defined in the [HTTP/HTTPS Configuration](doc:httphttps-configuration) section, unless access is restricted to dedicated ports using the corresponding configuration parameter (see below).
+
+Certificate authentication support is only available on the **HTTPS With Certificate Authentication Ports** configured in [HTTP/HTTPS Configuration](doc:httphttps-configuration) section.
```
Comment (mattdibi) [inline docs/core-services/rest-service.md:9]: Broken link, is this expected?

File: docs/core-services/nvidia-triton-server-inference-engine.md [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+- **Inference Models**: a comma-separated list of inference model names that the server will load. The models have to be already present in the filesystem where the server is running. This option simply tells the server to load the given models from a local or remote repository.
+- **Local backends path**: Specify the path on the filesystem where the backends are stored.
+- **Optional configuration for the local backends**: A semi-colon separated list of configuration for the backends. i.e. tensorflow,version=2;tensorflow,allow-soft-placement=false 
+- **Timeout (in seconds) for time consuming tasks**: Timeout (in seconds) for time consuming tasks like server startup, shutdown or model load. If the task exceeds the timeout, the operation will be terminated with an error.
+- **Max. GRPC message size (bytes)**: this field controls the maximum allowed size for the GRPC calls to the server instance.
+
+> **Note**: Pay attention on the ports used for communicating with the Triton Server. The default ports are the 8000-8002, but these are tipically used by Kura for debug purposes.
+
+
```
Comment (mattdibi) [inline docs/core-services/nvidia-triton-server-inference-engine.md:98]: There should be a screenshot here... did you forget it or did you do this by choice?

File: docs/core-services/nvidia-triton-server-inference-engine.md [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+- **CPUs**: Specify how many CPUs the Triton container can use. Decimal values are allowed, so if set to 1.5, the container will use at most one and a half cpu resource.
+- **GPUs**: Specify how many Nvidia GPUs the Triton container can use. Allowed values are 'all' or an integer number. If there's no Nvidia GPU installed, leave the field empty.
+- **Timeout (in seconds) for time consuming tasks**: Timeout (in seconds) for time consuming tasks like server startup, shutdown or model load. If the task exceeds the timeout, the operation will be terminated with an error.
+- **Max. GRPC message size (bytes)**: this field controls the maximum allowed size for the GRPC calls to the server instance.
+
+!!! note
+    Pay attention on the ports used for communicating with the Triton Server. The default ports are the 8000-8002, but these are typically used by Kura for debug purposes.
+
+
```
Comment (mattdibi) [inline docs/core-services/nvidia-triton-server-inference-engine.md:127]: There should be a screenshot here... did you forget it or did you do this by choice?

File: docs/core-services/nvidia-triton-server-inference-engine.md [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+
+## Triton Server Service component [deprecated since 5.2.0]
+
+The Kura Triton Server component is the implementation of the inference engine APIs and provides methods for interacting with a local or remote Nvidia™ Triton Server. As presented below, the component enables the user to configure a local server running on the gateway or to communicate to an external server to load specific models.
+
```
Comment (mattdibi) [inline docs/core-services/nvidia-triton-server-inference-engine.md:132]: There should be a screenshot here... did you forget it or did you do this by choice?

**Review by mattdibi — CHANGES_REQUESTED** [review]
There are a few broken links here and there, are these expected?

Also a few missing images in the Nvidia Triton Server section, is this expected?

Also the `Cloudcat` section from the old documentation is missing, is this because it is obsolete?

I added a few suggestion for rendering issues.

Overall LGTM 👍

**Comment by marcellorinaldo** [comment]
> There are a few broken links here and there, are these expected?

Some sections do not exist yet, whereas I missed some other links.

> Also a few missing images in the Nvidia Triton Server section, is this expected?

I tried to reduce the number of images to the bare minimum. Keep images up to date is more difficult than text. By the way, there is a lot of redundancy for the Nvidia Trition server documentation.

> Also the Cloudcat section from the old documentation is missing, is this because it is obsolete?

No, will add later.

File: docs/core-services/configuration-service-rest-v1.md (hunk 11 above)
Comment (mattdibi) [reply docs/core-services/configuration-service-rest-v1.md]: LOL, do not merge the suggestion as is: due to the ``` Github got confused

**Comment by mattdibi** [comment]
@marcellorinaldo 

> Yes, those sections do not exist yet.

As I did for my branches. LGTM was just double checking

> I tried to reduce the number of images to the bare minimum. Keep images up to date is more difficult than text. Keep images up to date is more difficult than text. By the way, there is a lot of redundancy for the Nvidia Trition server documentation.

Fair point.

> No, will add later.

👍

File: docs/core-services/rest-service.md (hunk 21 above)
Comment (marcellorinaldo) [reply docs/core-services/rest-service.md:7]: Section not yet ported.

File: docs/core-services/rest-service.md (hunk 23 above)
Comment (marcellorinaldo) [reply docs/core-services/rest-service.md:9]: Section not yet ported.

**Pull request merged**